### ISSUE TITLE
Few issues / questions

### ISSUE THREAD

**Issue by robbertkl** [just-containers/s6-overlay#107]
Hi,

Just started using s6-overlay (coming from supervisord) and love it so far. I do have a few issues / questions. Please let me know if I should submit separate issues.

1. On [this line of init-stage3](https://github.com/just-containers/s6-overlay/blob/a0d61a94fe1126348efb6396175bac63b2644f96/builder/overlay-rootfs/etc/s6/init/init-stage3#L48), `s6-nuke` is called with both `-t` and `-h`. This causes both `HUP` and `TERM` to be sent to all processes, even though the informative log message mentions only `TERM`. Shouldn't this be `-t` only? The `HUP` causes my process to reload their config, just before getting killed anyway. If there is valid reason for this (which I'd love to learn), I would suggest changing the log message.

2. Also in [init-stage3](https://github.com/just-containers/s6-overlay/blob/a0d61a94fe1126348efb6396175bac63b2644f96/builder/overlay-rootfs/etc/s6/init/init-stage3#L54), this stage is sleeping (by default) for 5 seconds, before sending `KILL`. It seems to do this every time, even if there are no more processes to "nuke". Even if I have a container running services that terminate quickly, `docker stop` (or restarts) will still take 5 seconds, which seems unnecessary. Any ideas / solutions for this?

3. This might be more of an issue with s6-svscan instead of s6-overlay. With supervisord (running as pid 1), sending a `HUP` to it (or `docker kill -s HUP`), would cause the `HUP` to be "forwarded" to all running services. This in turn would make all services reload their configs, as is a common convention for many daemons. With s6-overlay, this does not happen, instead causing the container to (cleanly) exit. Is there a way to forward signals like supervisord does? Might be useful for other signals as well, like `USR1` or `USR2`. Perhaps this is possible using `s6-svscan -s`, but if I understand the documentation correctly, this means we have to handle (i.e. properly deal with) all 6 signals in `.s6-svscan/` scripts, or else signals we be ignored. Any suggestions?

Thanks!

**Comment by glerchundi** [comment]
hi @robbertkl, first of all thanks for getting that deep in the code/behaviour!

> On this line of init-stage3, s6-nuke is called with both -t and -h. This causes both HUP and TERM to be sent to all processes, even though the informative log message mentions only TERM. Shouldn't this be -t only? The HUP causes my process to reload their config, just before getting killed anyway. If there is valid reason for this (which I'd love to learn), I would suggest changing the log message.

I followed the example @skarnet proposed in his `s6` package. As he knows much more than me about init systems, I would like to have his opinion. Perhaps just sending TERM it's ok.

> Also in init-stage3, this stage is sleeping (by default) for 5 seconds, before sending KILL. It seems to do this every time, even if there are no more processes to "nuke". Even if I have a container running services that terminate quickly, docker stop (or restarts) will still take 5 seconds, which seems unnecessary. Any ideas / solutions for this?

It's just to wait your services/processes terminate gracefully, it's the default behaviour but you can of course adapt to your needs. Take your services usually finish in less than 300ms hence you can set/override `S6_KILL_GRACETIME` with 300.

Is this ok for you?

> This might be more of an issue with s6-svscan instead of s6-overlay. With supervisord (running as pid 1), sending a HUP to it (or docker kill -s HUP), would cause the HUP to be "forwarded" to all running services. This in turn would make all services reload their configs, as is a common convention for many daemons. With s6-overlay, this does not happen, instead causing the container to (cleanly) exit. Is there a way to forward signals like supervisord does? Might be useful for other signals as well, like USR1 or USR2. Perhaps this is possible using s6-svscan -s, but if I understand the documentation correctly, this means we have to handle (i.e. properly deal with) all 6 signals in .s6-svscan/ scripts, or else signals we be ignored. Any suggestions?

This is a really interesting question. I haven't found in your position but I think it is an interesting use case. As you already guessed, there is a better forum for this issue which is the supervision mailing list (http://skarnet.org/lists.html). As far as I know I think there isn't a clean way to propagate a signal to supervised processes. It would be cool to have something like `s6-svscanctl -hc` HUP (-h) to supervised processes and continue (-c) without bringing down the supervisor itself.

/cc @skarnet

**Comment by jprjr** [comment]
@glerchundi I think the `HUP` is because the `logutil-service` runs `s6-log` with the `-p` switch, so `s6-log` only exits on a `HUP`, not on a `TERM` https://github.com/just-containers/s6-overlay/blob/master/builder/overlay-rootfs/usr/bin/logutil-service-main#L11

I think on a usual install or VM you'd want that "don't exit on `TERM`" behavior, but I'm wondering if we're OK just letting it exit on a `TERM` in our project. I'm not sure though.

**Comment by robbertkl** [comment]
As for # 1, I'm now running a modified version of `init-stage3` with just `-t`. It works fine so far in my situation, but I can't say I've tested it extensively.

With the original version (both `-t` and `-h`), I ran into another issue: I run a container with php-fpm, then refresh a page a couple of times (spinning up multiple php processes in the pool), followed by a control-C (`INT`) in the docker process. Following these steps, _most_ of the time, this explodes my log with:

```
[23-Nov-2015 15:49:27] NOTICE: [pool www] child 218 exited with code 0 after 0.013069 seconds from start
[23-Nov-2015 15:49:27] NOTICE: [pool www] child 219 started
[23-Nov-2015 15:49:27] NOTICE: [pool www] child 217 exited with code 0 after 0.015270 seconds from start
[23-Nov-2015 15:49:27] NOTICE: [pool www] child 220 started
[23-Nov-2015 15:49:27] NOTICE: [pool www] child 219 exited with code 0 after 0.015375 seconds from start
[23-Nov-2015 15:49:27] NOTICE: [pool www] child 221 started
...
(continuing for 5 seconds until KILL signal, generating hundreds of these lines)
```

With my modified version (only `-t`), I tried the same steps multiple times, but could not reproduce the same issue. The above is just to show a real-life scenario where the `HUP`+`TERM` seems to be causing issues.

**Comment by robbertkl** [comment]
For # 2, thanks @glerchundi for your suggestion to tweak `S6_KILL_GRACETIME`. I will definitely lower this on some containers, although I'd rather not risk having this too short causing premature `KILL`s. What would be the wisest way to loop and check every 100ms if there are any processes remaining? Are pid 1 and the `s6-nuke` process the only ones expected to still be alive? Is there any tool for the job in the s6 toolbelt? Thanks!

**Comment by glerchundi** [comment]
> @glerchundi I think the HUP is because the logutil-service runs s6-log with the -p switch, so s6-log only exits on a HUP, not on a TERM https://github.com/just-containers/s6-overlay/blob/master/builder/overlay-rootfs/usr/bin/logutil-service-main#L11

good catch!

> Another option is to send TERM with s6-nuke and explicitly send a HUP to the uncaught-logs service.

Yep, perhaps it's the way to go.

@robbertkl that is really awful :(, @skarnet & @jprjr what do you think if we remove both parameters? `-p` in `s6-log` and `-h` in `s6-nuke` if I have to be honest I really don't understand which are the counterparts of not using `-p` in s6-log if we're really bringing our container down. From skarnet's documentation: 

> -p : protect against SIGTERM. Do not exit on receipt of a SIGTERM; only exit on a SIGHUP or when reading EOF on stdin. This is useful for a logger that you really do not want to lose even if automated administration (e.g. the downing of a supervision tree) would kill it.

**Comment by skarnet** [comment]
1. The reason for the `-h` use of s6-nuke is that some processes will ignore a SIGTERM and only finish on a SIGHUP: for instance, interactive shells do that.
 The catch-all logger is not important here: it will exit on its own a few seconds later anyway (with the `fdclose 1 fdclose 2` statement). It does not matter if it is killed here or not. So you can remove the `-h` if you're certain that you don't have anything else that would resist a SIGTERM.

2. `S6_KILL_GRACETIME` is indeed the way the overlay allows you to configure it.
There's unfortunately no current way to cleanly wait for all children with a timeout. Maybe later, with cgroups support?
 If the only services you want to give grace time to are your supervised services, then you should be able to whip up something with `s6-svwait -d` - that's what it's for. But if you have subprocesses the supervisors don't know about, then there's no general solution. It's a problem that all init systems have had since the beginning of Unix.

3. `s6-svscan -s`, recently added, is indeed the correct way to handle your use case and do specific things with signals. Maybe the overlay could be changed to use it and provide standard default actions (as shown in the `examples/` subdirectory of the s6 tarball), which you could then override with your own scripts? Or, if @glerchundi wants to make things really user-friendly, he could test for an environment variable in `.s6-svscan/SIGHUP` and send a SIGHUP to all services if that variable is defined. :-)

**Comment by robbertkl** [comment]
Very clear @skarnet, thanks!

For # 1, would a solution be to change the order of signals in `s6-nuke`, when multiple flags are given? Currently, after parsing the command line arguments, it does first `HUP`, then `TERM`, then `KILL` (if the corresponding flags are passed of course). Wouldn't it make sense to reverse this order? If `TERM` is sent before `KILL` in order to give the process a chance to terminate cleanly, it's pretty much useless if `KILL` is sent immediately after.

Reversing the order and passing both `-t` and `-h` would send daemons a `TERM` first, letting them terminate. If the daemon is well-written, a `HUP` during the clean shutdown process should not interfere. As for interactive shells (or other processes ignoring `TERM`), the `HUP` would follow right after.

Just an idea. I don't know the reasons behind the current order in `s6-nuke`.

**Comment by skarnet** [comment]
The position of the SIGKILL doesn't matter at all. Even if the targeted processes start handling a SIGHUP, a SIGTERM or anything else, they will receive the SIGKILL immediately afterwards and will die anyway.

The relative position of SIGTERM and SIGHUP *may* be relevant if you have a loop that synchronously catches and handles signals, as is done for instance with a self-pipe; but even then, signal processing should never start an action that takes significant time before going back to processing the next signal - if it does, it is a bug in the application design. If php-fpm is normally supposed to exit on a SIGTERM, but does not exit immediately or semi-immediately when it receives a SIGHUP right before a SIGTERM, then it is a bug in php-fpm and that's what should be fixed. In the meantime, keep your `s6-nuke -t` without the `-h` since it's working for you.

It's not that I don't want to change the order in which the signals are sent. It's that it should really not make any difference and I don't want people to start relying on that order; and if I change it to send the SIGTERM first, it may fix your use case, but the exact symmetrical problem (SIGHUP not working properly after a caught SIGTERM) may appear instead in buggy applications, and the change will only displace the issue, not fix it.

**Comment by glerchundi** [comment]
> The reason for the -h use of s6-nuke is that some processes will ignore a SIGTERM and only finish on a SIGHUP: for instance, interactive shells do that.
> The catch-all logger is not important here: it will exit on its own a few seconds later anyway (with the fdclose 1 fdclose 2 statement). It does not matter if it is killed here or not. So you can remove the -h if you're certain that you don't have anything else that would resist a SIGTERM.

So for concluding, and with your comments in our minds, I think it's ok to send just SIGTERM to gracefully stop processes even knowing that we'll break backward compatibility (and remove `-p` from `s6-log` of course). I usually follow this pattern SIGHUP for hot-reconfiguration and SIGTERM for graceful shutdown and multiple known software do exactly the same, nginx, haproxy, mariadb, etc...

Whoever has any objections, speak now or forever hold your peace :-)

> s6-svscan -s, recently added, is indeed the correct way to handle your use case and do specific things with signals. Maybe the overlay could be changed to use it and provide standard default actions (as shown in the examples/ subdirectory of the s6 tarball), which you could then override with your own scripts? Or, if @glerchundi wants to make things really user-friendly, he could test for an environment variable in .s6-svscan/SIGHUP and send a SIGHUP to all services if that variable is defined. :-)

Good idea and I'll do pretty soon, `S6_*` envs are cheap and I think this customization is very helpful for a lot of kind of situations. Although we need to decide which envs should be used.

**Comment by robbertkl** [comment]
@skarnet I agree, let's leave s6-nuke as is.
@glerchundi Thanks for the change (proposal). And having flexibility in the way how signals are handled or forwarded would be awesome!

**Comment by skarnet** [comment]
You don't need to remove the `-p` option to `s6-log` in the catch-all logger. Again, it won't make any difference whether the logger dies on the SIGTERM or not. If you find it does make a difference, it's a bug and we will fix it.

**Comment by glerchundi** [comment]
ok, thanks.

**Comment by glerchundi** [comment]
this seems answered, i'll implemented SIG* signals handling in one month or so.
